lib/puppet/parser/functions/new.rb

```diff
@@ -0,0 +1,530 @@
+allows different formatting depending on type.
+
+**Example:** Positive Integers in Hexadecimal prefixed with '0x', negative in Decimal
+
+```puppet
+$format_map = {
+  Integer[default, 0] => "%d",
+  Integer[1, default] => "%#x"
+}
+String("-1", $format_map)  # produces '-1'
+String("10", $format_map)  # produces '0xa'
+```
+
+A format is specified on the form:
+
+```
+%[Flags][Width][.Precision]Format
+```
+
+`Width` is the number of characters into which the value should be fitted. This allocated space is
+padded if value is shorter. By default it is space padded, and the flag `0` will cause padding with `0`
+for numerical formats.
+
+`Precision` is the number of fractional digits to show for floating point, and the maximum characters
+included in a string format.
+
+Note that all data type supports the formats `s` and `p` with the meaning "default string representation" and
+"default programmatic string representation" (which for example means that a String is quoted in 'p' format).
+
+### Signatures of String conversion
+
+```puppet
+type Format = Pattern[/^%([\s\+\-#0\[\{<\(\|]*)([1-9][0-9]*)?(?:\.([0-9]+))?([a-zA-Z])/]
+type ContainerFormat = Struct[{
+  format         => Optional[String],
+  separator      => Optional[String],
+  separator2     => Optional[String],
+  string_formats => Hash[Type, Format]
+  }]
+type TypeMap = Hash[Type, Variant[Format, ContainerFormat]]
+type Formats = Variant[Default, String[1], TypeMap]
+
+function String.new(
+  Any $value,
+  Formats $string_formats
+)
+```
+
+Where:
+
+* `separator` is the string used to separate entries in an array, or hash (extra space should not be included at
+  the end), defaults to `","`
+* `separator2` is the separator between key and value in a hash entry (space padding should be included as
+  wanted), defaults to `" => "`.
+* `string_formats` is a data type to format map for values contained in arrays and hashes - defaults to `{Any => "%p"}`. Note that
+  these nested formats are not applicable to data types that are containers; they are always formatted as per the top level
+  format specification.
+
+**Example:** Simple Conversion to String (using defaults)
+
+```puppet
+$str = String(10)      # produces '10'
+$str = String([10])    # produces '["10"]'
+```
+
+**Example:** Simple Conversion to String specifying the format for the given value directly
+
+```puppet
+$str = String(10, "%#x")    # produces '0x10'
+$str = String([10], "%(a")  # produces '("10")'
+```
+
+**Example:** Specifying type for values contained in an array
+
+```puppet
+$formats = { 
+  Array => {
+    format => '%(a',
+    string_formats => { Integer => '%#x' }
+  }
+}
+$str = String([1,2,3], $formats) # produces '(0x1, 0x2, 0x3)'
+```
+
+The given formats are merged with the default formats, and matching of values to convert against format is based on
+the specificity of the mapped type; for example, different formats can be used for short and long arrays.
+
+### Integer to String
```

Comment:
In lieu of making a lot of small line comments, from 258-386, the format descriptions need some consistency. Each description should start with capital letter, and end with a period. This includes the shorter descriptions, so `same as d` should be `Same as d.`